File: Source/Engine/Graphics/Textures/GPUSamplerDescription.cpp
// Copyright (c) 2012-2020 Wojciech Figat. All rights reserved.

#include "GPUSamplerDescription.h"
#include "Engine/Core/Types/String.h"

void GPUSamplerDescription::Clear()
{
    Filter = GPUSamplerFilter::Point;
    AddressU = GPUSamplerAddressMode::Wrap;
    AddressV = GPUSamplerAddressMode::Wrap;
    AddressW = GPUSamplerAddressMode::Wrap;
    MipBias = 0;
    MaxAnisotropy = 0;
    MinMipLevel = 0;
    MaxMipLevel = MAX_float;
    BorderColor = GPUSamplerBorderColor::TransparentBlack;
    SamplerComparisonFunction = GPUSamplerCompareFunction::Never;
}

bool GPUSamplerDescription::Equals(const GPUSamplerDescription& other) const
{
    return Filter == other.Filter
            && AddressU == other.AddressU
            && AddressV == other.AddressV
            && AddressW == other.AddressW
            && MipBias == other.MipBias
            && MaxAnisotropy == other.MaxAnisotropy
            && MinMipLevel == other.MinMipLevel
            && MaxMipLevel == other.MaxMipLevel
            && BorderColor == other.BorderColor
            && SamplerComparisonFunction == other.SamplerComparisonFunction;
}

String GPUSamplerDescription::ToString() const
{
    return String::Format(TEXT("Filter: {}, Address: {}x{}x{}, MipBias: {}, MaxAnisotropy: {}, MinMipLevel: {}, MaxMipLevel: {}, BorderColor: {}, SamplerComparisonFunction: {}"),
                          (int32)Filter,
                          (int32)AddressU,
                          (int32)AddressV,
                          (int32)AddressW,
                          MipBias,
                          MaxAnisotropy,
                          MinMipLevel,
                          MaxMipLevel,
                          (int32)BorderColor,
                          (int32)SamplerComparisonFunction);
}

uint32 GetHash(const GPUSamplerDescription& key)
{
    uint32 hashCode = (uint32)key.Filter;
    hashCode = (hashCode * 397) ^ (uint32)key.AddressU;
    hashCode = (hashCode * 397) ^ (uint32)key.AddressV;
    hashCode = (hashCode * 397) ^ (uint32)key.AddressW;
    hashCode = (hashCode * 397) ^ (uint32)key.MipBias;
    hashCode = (hashCode * 397) ^ (uint32)key.MaxAnisotropy;
    hashCode = (hashCode * 397) ^ (uint32)key.MinMipLevel;
    hashCode = (hashCode * 397) ^ (uint32)key.MaxMipLevel;
    hashCode = (hashCode * 397) ^ (uint32)key.BorderColor;
    hashCode = (hashCode * 397) ^ (uint32)key.SamplerComparisonFunction;
    return hashCode;
}

File: Source/Engine/Graphics/Textures/GPUSamplerDescription.cs
// Copyright (c) Wojciech Figat. All rights reserved.

using System;

namespace FlaxEngine
{
    partial struct GPUSamplerDescription : IEquatable<GPUSamplerDescription>
    {
        /// <summary>
        /// Clears description.
        /// </summary>
        public void Clear()
        {
            this = new GPUSamplerDescription();
            MaxMipLevel = float.MaxValue;
        }

        /// <summary>
        /// Creates a new <see cref="GPUSamplerDescription" /> with default settings.
        /// </summary>
        /// <param name="filter">The filtering method.</param>
        /// <param name="addressMode">The addressing mode.</param>
        /// <returns>A new instance of <see cref="GPUSamplerDescription" /> class.</returns>
        public static GPUSamplerDescription New(GPUSamplerFilter filter = GPUSamplerFilter.Point, GPUSamplerAddressMode addressMode = GPUSamplerAddressMode.Wrap)
        {
            return new GPUSamplerDescription
            {
                Filter = filter,
                AddressU = addressMode,
                AddressV = addressMode,
                AddressW = addressMode,
                MaxMipLevel = float.MaxValue,
            };
        }

        /// <inheritdoc />
        public bool Equals(GPUSamplerDescription other)
        {
            return Filter == other.Filter &&
                   AddressU == other.AddressU &&
                   AddressV == other.AddressV &&
                   AddressW == other.AddressW &&
                   Mathf.NearEqual(MipBias, other.MipBias) &&
                   Mathf.NearEqual(MinMipLevel, other.MinMipLevel) &&
                   Mathf.NearEqual(MaxMipLevel, other.MaxMipLevel) &&
                   MaxAnisotropy == other.MaxAnisotropy &&
                   BorderColor == other.BorderColor &&
                   ComparisonFunction == other.ComparisonFunction;
        }

        /// <inheritdoc />
        public override bool Equals(object obj)
        {
            return obj is GPUSamplerDescription other && Equals(other);
        }

        /// <inheritdoc />
        public override int GetHashCode()
        {
            unchecked
            {
                var hashCode = (int)Filter;
                hashCode = (hashCode * 397) ^ (int)AddressU;
                hashCode = (hashCode * 397) ^ (int)AddressV;
                hashCode = (hashCode * 397) ^ (int)AddressW;
                hashCode = (hashCode * 397) ^ (int)MipBias;
                hashCode = (hashCode * 397) ^ (int)MinMipLevel;
                hashCode = (hashCode * 397) ^ (int)MaxMipLevel;
                hashCode = (hashCode * 397) ^ MaxAnisotropy;
                hashCode = (hashCode * 397) ^ (int)BorderColor;
                hashCode = (hashCode * 397) ^ (int)ComparisonFunction;
                return hashCode;
            }
        }
    }
}
